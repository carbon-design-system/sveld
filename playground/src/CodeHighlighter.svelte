<script context="module">
  import json from "svelte-highlight/languages/json";
  import typescript from "svelte-highlight/languages/typescript";
  import markdown from "svelte-highlight/languages/markdown";

  const LANG = {
    json,
    typescript,
    markdown,
  };
</script>

<script>
  export let code = "";
  export let language = "typescript";
  export let noWrap = false;

  import { CopyButton } from "carbon-components-svelte";
  import Highlight from "svelte-highlight";
  import "svelte-highlight/styles/zenburn.css";
</script>

<div class="code-highlighter" class:noWrap>
  <div>
    <CopyButton text={code} />
  </div>
  <Highlight language={LANG[language]} {code} />
</div>

<style>
  .code-highlighter {
    position: relative;
    display: flex;
    flex: 1;
    width: 100%;
  }

  .code-highlighter div {
    position: absolute;
    top: 0.5rem;
    right: 1.5rem;
  }

  :global(code.hljs) {
    background: var(--cds-ui-01); /** TODO: use token */
    font-family: var(--cds-code-02-font-family);
    font-size: var(--cds-code-02-font-size);
    font-weight: var(--cds-code-02-font-weight);
    letter-spacing: var(--cds-code-02-letter-spacing);
    line-height: var(--cds-code-02-line-height);
    cursor: text;
  }

  :global(.code-highlighter:not(.noWrap) code.hljs) {
    white-space: pre-wrap;
  }

  :global(pre code.hljs) {
    padding: 1rem;
  }
</style>
