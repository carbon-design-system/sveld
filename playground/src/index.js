import "carbon-components-svelte/css/all.css";
import App from "./App.svelte";

new App({ target: document.body });