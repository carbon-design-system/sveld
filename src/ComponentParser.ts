import { compile, walk } from "svelte/compiler";
import * as commentParser from "comment-parser";
import { Ast, TemplateNode, Var } from "svelte/types/compiler/interfaces";
import { getElementByTag } from "./element-tag-map";
import { Node } from "estree-walker";
import type { VariableDeclaration } from "estree";

interface CompiledSvelteCode {
  vars: Var[];
  ast: Ast;
}

interface ComponentParserDiagnostics {
  moduleName: string;
  filePath: string;
}

interface ComponentParserOptions {
  verbose?: boolean;
}

type ComponentPropName = string;

interface ComponentProp {
  name: string;
  kind: "let" | "const" | "function";
  constant: boolean;
  type?: string;
  value?: any;
  description?: string;
  isFunction: boolean;
  isFunctionDeclaration: boolean;
  reactive: boolean;
}

const DEFAULT_SLOT_NAME = "__default__";

type ComponentSlotName = typeof DEFAULT_SLOT_NAME | string;

interface ComponentSlot {
  name?: string;
  default: boolean;
  fallback?: string;
  slot_props?: string;
}

interface SlotPropValue {
  value?: string;
  replace: boolean;
}

type SlotProps = Record<string, SlotPropValue>;

type ComponentEventName = string;

interface ForwardedEvent {
  type: "forwarded";
  name: string;
  element: ComponentInlineElement | ComponentElement;
}

interface DispatchedEvent {
  type: "dispatched";
  name: string;
  detail?: any;
}

type ComponentEvent = ForwardedEvent | DispatchedEvent;

type TypeDefName = string;

interface TypeDef extends Pick<commentParser.Tag, "type" | "name"> {
  description?: string;
  ts: string;
}

interface ComponentInlineElement {
  type: "InlineComponent";
  name: string;
}

interface ComponentElement {
  type: "Element";
  name: string;
}

type RestProps = undefined | ComponentInlineElement | ComponentElement;

interface Extends {
  interface: string;
  import: string;
}

interface ComponentPropBindings {
  elements: string[];
}

export interface ParsedComponent {
  props: ComponentProp[];
  slots: ComponentSlot[];
  events: ComponentEvent[];
  typedefs: TypeDef[];
  rest_props: RestProps;
  extends?: Extends;
  componentComment?: string;
}

export default class ComponentParser {
  private options?: ComponentParserOptions;
  private source?: string;
  private compiled?: CompiledSvelteCode;
  private rest_props?: RestProps;
  private extends?: Extends;
  private componentComment?: string;
  private readonly reactive_vars: Set<string> = new Set();
  private readonly vars: Set<VariableDeclaration> = new Set();
  private readonly props: Map<ComponentPropName, ComponentProp> = new Map();
  private readonly slots: Map<ComponentSlotName, ComponentSlot> = new Map();
  private readonly events: Map<ComponentEventName, ComponentEvent> = new Map();
  private readonly typedefs: Map<TypeDefName, TypeDef> = new Map();
  private readonly bindings: Map<ComponentPropName, ComponentPropBindings> = new Map();

  constructor(options?: ComponentParserOptions) {
    this.options = options;
  }

  private static mapToArray(map: Map<any, any>) {
    return Array.from(map, ([key, value]) => value);
  }

  private static assignValue(value?: "" | string) {
    return value === undefined || value === "" ? undefined : value;
  }

  private static formatComment(comment: string) {
    let formatted_comment = comment;

    if (!formatted_comment.startsWith("/*")) {
      formatted_comment = "/*" + formatted_comment;
    }

    if (!formatted_comment.endsWith("*/")) {
      formatted_comment += "*/";
    }

    return formatted_comment;
  }

  private sourceAtPos(start: number, end: number) {
    return this.source?.slice(start, end);
  }

  private collectReactiveVars() {
    this.compiled?.vars
      .filter(({ reassigned, writable }) => reassigned && writable)
      .forEach(({ name }) => this.reactive_vars.add(name));
  }

  private addProp(prop_name: string, data: ComponentProp) {
    if (ComponentParser.assignValue(prop_name) === undefined) return;

    if (this.props.has(prop_name)) {
      const existing_slot = this.props.get(prop_name)!;

      this.props.set(prop_name, {
        ...existing_slot,
        ...data,
      });
    } else {
      this.props.set(prop_name, data);
    }
  }

  private aliasType(type: any) {
    if (type === "*") return "any";
    return type;
  }

  private addSlot(slot_name?: string, slot_props?: string, slot_fallback?: string) {
    const default_slot = slot_name === undefined || slot_name === "";
    const name: ComponentSlotName = default_slot ? DEFAULT_SLOT_NAME : slot_name!;
    const fallback = ComponentParser.assignValue(slot_fallback);
    const props = ComponentParser.assignValue(slot_props);

    if (this.slots.has(name)) {
      const existing_slot = this.slots.get(name)!;

      this.slots.set(name, {
        ...existing_slot,
        fallback,
        slot_props: existing_slot.slot_props === undefined ? props : existing_slot.slot_props,
      });
    } else {
      this.slots.set(name, {
        name,
        default: default_slot,
        fallback,
        slot_props,
      });
    }
  }

  private addDispatchedEvent(name?: string, detail?: string) {
    if (name === undefined) return;
    if (this.events.has(name)) {
      const existing_event = this.events.get(name) as DispatchedEvent;

      this.events.set(name, {
        ...existing_event,
        detail: existing_event.detail === undefined ? ComponentParser.assignValue(detail) : existing_event.detail,
      });
    } else {
      this.events.set(name, {
        type: "dispatched",
        name,
        detail: ComponentParser.assignValue(detail),
      });
    }
  }

  private parseCustomTypes() {
    commentParser(this.source!).forEach(({ tags }) => {
      tags.forEach(({ tag, type: tagType, name, description }) => {
        const type = this.aliasType(tagType);

        switch (tag) {
          case "extends":
            this.extends = {
              interface: name,
              import: type,
            };
            break;
          case "restProps":
            this.rest_props = {
              type: "Element",
              name: type,
            };
            break;
          case "slot":
            this.addSlot(name, type);
            break;
          case "event":
            this.addDispatchedEvent(name, type);
            break;
          case "typedef":
            this.typedefs.set(name, {
              type,
              name,
              description: ComponentParser.assignValue(description),
              ts: /(\}|\};)$/.test(type) ? `interface ${name} ${type}` : `type ${name} = ${type}`,
            });
            break;
        }
      });
    });
  }

  public cleanup() {
    this.source = undefined;
    this.compiled = undefined;
    this.rest_props = undefined;
    this.extends = undefined;
    this.componentComment = undefined;
    this.reactive_vars.clear();
    this.props.clear();
    this.slots.clear();
    this.events.clear();
    this.typedefs.clear();
    this.bindings.clear();
  }

  public parseSvelteComponent(source: string, diagnostics: ComponentParserDiagnostics): ParsedComponent {
    if (this.options?.verbose) {
      process.stdout.write(`[parsing] "${diagnostics.moduleName}" ${diagnostics.filePath}\n`);
    }

    this.cleanup();
    this.source = source;
    this.compiled = compile(source);
    this.collectReactiveVars();
    this.parseCustomTypes();

    let dispatcher_name: undefined | string = undefined;
    let callees: { name: string; arguments: any }[] = [];

    walk(this.compiled.ast as unknown as Node, {
      enter: (node, parent, prop) => {
        if (node.type === "CallExpression") {
          if (node.callee.name === "createEventDispatcher") {
            dispatcher_name = parent?.id.name;
          }

          callees.push({
            name: node.callee.name,
            arguments: node.arguments,
          });
        }

        if (node.type === "Spread" && node?.expression.name === "$$restProps") {
          if (this.rest_props === undefined && (parent?.type === "InlineComponent" || parent?.type === "Element")) {
            this.rest_props = {
              type: parent.type,
              name: parent.name,
            };
          }
        }

        if (node.type === "VariableDeclaration") {
          this.vars.add(node as unknown as VariableDeclaration);
        }

        if (node.type === "ExportNamedDeclaration") {
          // Handle renamed exports
          let prop_name: string;
          if (node.declaration == null && node.specifiers[0]?.type === "ExportSpecifier") {
            const specifier = node.specifiers[0];
            const localName = specifier.local.name,
              exportedName = specifier.exported.name;
            let declaration: VariableDeclaration;
            // Search through all variable declarations for this variable
            //  Limitation: the variable must have been declared before the export
            this.vars.forEach((varDecl) => {
              if (varDecl.declarations.some((decl) => decl.id.type === "Identifier" && decl.id.name === localName)) {
                declaration = varDecl;
              }
            });
            node.declaration = declaration!;
            prop_name = exportedName;
          }

          const {
            type: declaration_type,
            id,
            init,
            body,
          } = node.declaration.declarations ? node.declaration.declarations[0] : node.declaration;

          prop_name ??= id.name;

          let value = undefined;
          let type = undefined;
          let kind = node.declaration.kind;
          let description = undefined;
          let isFunction = false;
          let isFunctionDeclaration = false;

          if (init != null) {
            if (
              init.type === "ObjectExpression" ||
              init.type === "BinaryExpression" ||
              init.type === "ArrayExpression" ||
              init.type === "ArrowFunctionExpression"
            ) {
              value = this.sourceAtPos(init.start, init.end)?.replace(/\n/g, " ");
              type = value;
              isFunction = init.type === "ArrowFunctionExpression";

              if (init.type === "BinaryExpression") {
                if (init?.left.type === "Literal" && typeof init?.left.value === "string") {
                  type = "string";
                }
              }
            } else {
              if (init.type === "UnaryExpression") {
                value = this.sourceAtPos(init.start, init.end);
                type = typeof init.argument?.value;
              } else {
                value = init.raw;
                type = init.value == null ? undefined : typeof init.value;
              }
            }
          }

          if (declaration_type === "FunctionDeclaration") {
            value = "() => " + this.sourceAtPos(body.start, body.end)?.replace(/\n/g, " ");
            type = "() => any";
            kind = "function";
            isFunction = true;
            isFunctionDeclaration = true;
          }

          if (node.leadingComments) {
            const last_comment = node.leadingComments[node.leadingComments.length - 1];
            const comment = commentParser(ComponentParser.formatComment(last_comment.value));
            const tag = comment[0]?.tags[comment[0]?.tags.length - 1];
            if (tag?.tag === "type") type = this.aliasType(tag.type);
            description = ComponentParser.assignValue(comment[0]?.description);
          }

          if (!description && this.typedefs.has(type)) {
            description = this.typedefs.get(type)!.description;
          }

          this.addProp(prop_name, {
            name: prop_name,
            kind,
            description,
            type,
            value,
            isFunction,
            isFunctionDeclaration,
            constant: kind === "const",
            reactive: this.reactive_vars.has(prop_name),
          });
        }

        if (node.type === "Comment") {
          let data: string = node?.data?.trim() ?? "";

          if (/^@component/.test(data)) {
            this.componentComment = data.replace(/^@component/, "");
          }
        }

        if (node.type === "Slot") {
          const slot_name = node.attributes.find((attr: any) => attr.name === "name")?.value[0].data;

          const slot_props = node.attributes
            .filter((attr: { name?: string }) => attr.name !== "name")
            .reduce((slot_props: SlotProps, { name, value }: { name: string; value?: any }) => {
              let slot_prop_value: SlotPropValue = {
                value: undefined,
                replace: false,
              };

              if (value === undefined) return {};

              if (value[0]) {
                const { type, expression, raw, start, end } = value[0];

                if (type === "Text") {
                  slot_prop_value.value = raw;
                } else if (type === "AttributeShorthand") {
                  slot_prop_value.value = expression.name;
                  slot_prop_value.replace = true;
                }

                if (expression) {
                  if (expression.type === "Literal") {
                    slot_prop_value.value = expression.value;
                  } else if (expression.type !== "Identifier") {
                    if (expression.type === "ObjectExpression" || expression.type === "TemplateLiteral") {
                      slot_prop_value.value = this.sourceAtPos(start + 1, end - 1);
                    } else {
                      slot_prop_value.value = this.sourceAtPos(start, end);
                    }
                  }
                }
              }

              return { ...slot_props, [name]: slot_prop_value };
            }, {});

          const fallback = (node.children as TemplateNode[])
            ?.map(({ start, end }) => this.sourceAtPos(start, end))
            .join("")
            .trim();

          this.addSlot(slot_name, JSON.stringify(slot_props, null, 2), fallback);
        }

        if (node.type === "EventHandler" && node.expression == null) {
          if (!this.events.has(node.name) && parent !== undefined) {
            this.events.set(node.name, {
              type: "forwarded",
              name: node.name,
              element: parent.name,
            });
          }
        }

        if (parent?.type === "Element" && node.type === "Binding" && node.name === "this") {
          const prop_name = node.expression.name;
          const element_name = parent.name;

          if (this.bindings.has(prop_name)) {
            const existing_bindings = this.bindings.get(prop_name)!;

            if (!existing_bindings.elements.includes(element_name)) {
              this.bindings.set(prop_name, {
                ...existing_bindings,
                elements: [...existing_bindings.elements, element_name],
              });
            }
          } else {
            this.bindings.set(prop_name, {
              elements: [element_name],
            });
          }
        }
      },
    });

    if (dispatcher_name !== undefined) {
      callees.forEach((callee) => {
        if (callee.name === dispatcher_name) {
          const event_name = callee.arguments[0]?.value;
          const event_detail = callee.arguments[1]?.value;

          this.addDispatchedEvent(event_name, event_detail);
        }
      });
    }

    return {
      props: ComponentParser.mapToArray(this.props).map((prop) => {
        if (this.bindings.has(prop.name)) {
          return {
            ...prop,
            type:
              "null | " +
              this.bindings
                .get(prop.name)!
                .elements.sort()
                .map((element) => getElementByTag(element))
                .join(" | "),
          };
        }

        return prop;
      }),
      slots: ComponentParser.mapToArray(this.slots)
        .map((slot) => {
          try {
            const slot_props: SlotProps = JSON.parse(slot.slot_props);
            const new_props: string[] = [];

            Object.keys(slot_props).forEach((key) => {
              if (slot_props[key].replace && slot_props[key].value !== undefined) {
                slot_props[key].value = this.props.get(slot_props[key].value!)?.type;
              }

              if (slot_props[key].value === undefined) slot_props[key].value = "any";
              new_props.push(`${key}: ${slot_props[key].value}`);
            });

            const formatted_slot_props = new_props.length === 0 ? "{}" : "{ " + new_props.join(", ") + " }";

            return { ...slot, slot_props: formatted_slot_props };
          } catch (e) {
            return slot;
          }
        })
        .sort((a, b) => {
          if (a.name! < b.name!) return -1;
          if (a.name! > b.name) return 1;
          return 0;
        }),
      events: ComponentParser.mapToArray(this.events),
      typedefs: ComponentParser.mapToArray(this.typedefs),
      rest_props: this.rest_props,
      extends: this.extends,
      componentComment: this.componentComment,
    };
  }
}
