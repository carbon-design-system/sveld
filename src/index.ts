export { default } from "./rollup-plugin";
export { default as ComponentParser } from "./ComponentParser";
export { cli } from "./cli";
export { sveld } from "./sveld";
