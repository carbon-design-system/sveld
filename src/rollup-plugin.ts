import * as fs from "fs";
import * as fsp from "fs/promises";
import * as path from "path";
import * as fg from "fast-glob";
import writeTsDefinitions, { WriteTsDefinitionsOptions } from "./writer/writer-ts-definitions";
import writeJson, { WriteJsonOptions } from "./writer/writer-json";
import writeMarkdown, { WriteMarkdownOptions } from "./writer/writer-markdown";
import ComponentParser, { ParsedComponent } from "./ComponentParser";
import { getSvelteEntry } from "./get-svelte-entry";
import { ParsedExports, parseExports } from "./parse-exports";
import { preprocess } from "svelte/compiler";
import { replace, typescript } from "svelte-preprocess";
import { normalizeSeparators } from "./path";

export interface PluginSveldOptions {
  glob?: boolean;
  types?: boolean;
  typesOptions?: Partial<Omit<WriteTsDefinitionsOptions, "inputDir">>;
  json?: boolean;
  jsonOptions?: Partial<Omit<WriteJsonOptions, "inputDir">>;
  markdown?: boolean;
  markdownOptions?: Partial<WriteMarkdownOptions>;
}

type ComponentModuleName = string;

export interface ComponentDocApi extends ParsedComponent {
  filePath: string;
  moduleName: string;
}

export type ComponentDocs = Map<ComponentModuleName, ComponentDocApi>;

export default function pluginSveld(opts?: PluginSveldOptions) {
  let result: GenerateBundleResult;
  let input: string | null;

  return {
    name: "plugin-sveld",
    buildStart() {
      input = getSvelteEntry();
    },
    async generateBundle() {
      if (input != null) {
        result = await generateBundle(input, opts?.glob === true);
      }
    },
    writeBundle() {
      if (input != null) writeOutput(result, opts || {}, input);
    },
  };
}

interface GenerateBundleResult {
  exports: ParsedExports;
  components: ComponentDocs;
}

export async function generateBundle(input: string, glob: boolean) {
  const dir = fs.lstatSync(input).isFile() ? path.dirname(input) : input;
  const entry = fs.readFileSync(input, "utf-8");
  const exports = parseExports(entry);

  if (glob) {
    fg.sync([`${dir}/**/*.svelte`]).forEach((file) => {
      const moduleName = path.parse(file).name.replace(/\-/g, "");
      let source = normalizeSeparators("./" + path.relative(dir, file));

      if (exports[moduleName]) {
        exports[moduleName].source = source;
      }
    });
  }

  const components: ComponentDocs = new Map();
  const parser = new ComponentParser();
  const exportEntries = Object.entries(exports);

  for (const [exportName, entry] of exportEntries) {
    const filePath = entry.source;
    const { ext, name } = path.parse(filePath);

    let moduleName = exportName;

    if (exportEntries.length === 1 && exportName === "default") {
      moduleName = name;
    }

    if (ext === ".svelte") {
      const source = await fsp.readFile(path.resolve(dir, filePath), "utf-8");
      const { code: processed } = await preprocess(source, [typescript(), replace([[/<style.+<\/style>/gims, ""]])], {
        filename: path.basename(filePath),
      });

      components.set(moduleName, {
        moduleName,
        filePath,
        ...parser.parseSvelteComponent(processed, {
          moduleName,
          filePath,
        }),
      });
    }
  }

  return {
    exports,
    components,
  };
}

export function writeOutput(result: GenerateBundleResult, opts: PluginSveldOptions, input: string) {
  const inputDir = path.dirname(input);

  if (opts?.types !== false) {
    writeTsDefinitions(result.components, {
      outDir: "types",
      preamble: "",
      ...opts?.typesOptions,
      exports: result.exports,
      inputDir,
    });
  }

  if (opts?.json) {
    writeJson(result.components, {
      outFile: "COMPONENT_API.json",
      ...opts?.jsonOptions,
      input,
      inputDir,
    });
  }

  if (opts?.markdown) {
    writeMarkdown(result.components, {
      outFile: path.join(process.cwd(), "COMPONENT_INDEX.md"),
      ...opts?.markdownOptions,
    });
  }
}
