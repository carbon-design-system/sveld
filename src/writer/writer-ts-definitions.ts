import * as path from "path";
import { convertSvelteExt, createExports } from "../create-exports";
import { ParsedExports } from "../parse-exports";
import { ComponentDocApi, ComponentDocs } from "../rollup-plugin";
import Writer from "./Writer";

const ANY_TYPE = "any";
const EMPTY_STR = "";

export function formatTsProps(props?: string) {
  if (props === undefined) return ANY_TYPE;
  return props + "\n";
}

export function getTypeDefs(def: Pick<ComponentDocApi, "typedefs">) {
  if (def.typedefs.length === 0) return EMPTY_STR;
  return def.typedefs.map((typedef) => `export ${typedef.ts}`).join("\n\n");
}

function clampKey(key: string) {
  if (/(\-|\s+|\:)/.test(key)) {
    return /(\"|\')/.test(key) ? key : `["${key}"]`;
  }

  return key;
}

function addCommentLine(value: any, returnValue?: any) {
  if (!value) return undefined;
  return `* ${returnValue || value}\n`;
}

function genPropDef(def: Pick<ComponentDocApi, "props" | "rest_props" | "moduleName" | "extends">) {
  const initial_props = def.props
    .filter((prop) => !prop.isFunctionDeclaration && prop.kind !== "const")
    .map((prop) => {
      let defaultValue = prop.value;

      if (typeof prop.value === "string") {
        defaultValue = prop.value.replace(/\s+/g, " ");
      }

      if (prop.value === undefined) {
        defaultValue = "undefined";
      }

      const prop_comments = [
        addCommentLine(prop.description?.replace(/\n/g, "\n* ")),
        addCommentLine(prop.constant, "@constant"),
        `* @default ${defaultValue}\n`,
      ]
        .filter(Boolean)
        .join("");

      let prop_value = prop.constant && !prop.isFunction ? prop.value : prop.type;

      return `
      ${prop_comments.length > 0 ? `/**\n${prop_comments}*/` : EMPTY_STR}
      ${prop.name}?: ${prop_value};`;
    });

  if (def.rest_props?.type === "Element") {
    const elements = def.rest_props?.name.split("|").map((element) => element.replace(/\s+/g, ""));

    if (elements.includes("a")) {
      initial_props.push(
        [
          "\n",
          `
          /**
           * SvelteKit attribute to enable data prefetching
           * if a link is hovered over or touched on mobile.
           * @see https://kit.svelte.dev/docs/a-options#sveltekit-prefetch
           * @default false
           */
           "sveltekit:prefetch"?: boolean;
          `,
          "\n",
          `
          /**
           * SvelteKit attribute to trigger a full page
           * reload after the link is clicked.
           * @see https://kit.svelte.dev/docs/a-options#sveltekit-reload
           * @default false
           */
           "sveltekit:reload"?: boolean;
          `,
          "\n",
          `
          /**
           * SvelteKit attribute to prevent scrolling
           * after the link is clicked.
           * @see https://kit.svelte.dev/docs/a-options#sveltekit-noscroll
           * @default false
           */
           "sveltekit:noscroll"?: boolean;
          `,
        ].join("\n")
      );
    }
  }

  const props = initial_props.join("\n");

  const props_name = `${def.moduleName}Props`;

  let prop_def = EMPTY_STR;

  if (def.rest_props?.type === "Element") {
    const extend_tag_map = def.rest_props.name
      .split("|")
      .map((name) => {
        const element = name.trim();

        if (element === "svg") {
          return "svelte.JSX.SVGAttributes<SVGSVGElement>";
        }

        return `svelte.JSX.HTMLAttributes<HTMLElementTagNameMap["${element}"]>`;
      })
      .join(",");

    prop_def = `
    export interface ${props_name} extends ${
      def.extends !== undefined ? `${def.extends.interface}, ` : ""
    }${extend_tag_map} {
      ${props}
    }
  `;
  } else {
    prop_def = `
    export interface ${props_name} ${def.extends !== undefined ? `extends ${def.extends.interface}` : ""} {
      ${props}
    }
  `;
  }

  return {
    props_name,
    prop_def,
  };
}

function genSlotDef(def: Pick<ComponentDocApi, "slots">) {
  return def.slots
    .map(({ name, slot_props, ...rest }) => {
      const key = rest.default ? "default" : clampKey(name!);
      return `${clampKey(key)}: ${formatTsProps(slot_props)};`;
    })
    .join("\n");
}

function genEventDef(def: Pick<ComponentDocApi, "events">) {
  const createDispatchedEvent = (detail: string = ANY_TYPE) => {
    if (/CustomEvent/.test(detail)) return detail;
    return `CustomEvent<${detail}>`;
  };
  return def.events
    .map((event) => {
      return `${clampKey(event.name)}: ${
        event.type === "dispatched" ? createDispatchedEvent(event.detail) : `WindowEventMap["${event.name}"]`
      };`;
    })
    .join("\n");
}

function genAccessors(def: Pick<ComponentDocApi, "props">) {
  return def.props
    .filter((prop) => prop.isFunctionDeclaration || prop.kind === "const")
    .map((prop) => {
      const prop_comments = [addCommentLine(prop.description?.replace(/\n/g, "\n* "))].filter(Boolean).join("");

      return `
    ${prop_comments.length > 0 ? `/**\n${prop_comments}*/` : EMPTY_STR}
    ${prop.name}: ${prop.type};`;
    })
    .join("\n");
}

function genImports(def: Pick<ComponentDocApi, "extends">) {
  if (def.extends === undefined) return "";
  return `import type { ${def.extends.interface} } from ${def.extends.import};`;
}

function genComponentComment(def: Pick<ComponentDocApi, "componentComment">) {
  if (!def.componentComment) return "";
  if (!/\n/.test(def.componentComment)) return `/** ${def.componentComment.trim()} */`;
  return `/*${def.componentComment
    .split("\n")
    .map((line) => `* ${line}`)
    .join("\n")}\n*/`;
}

function genModuleExports(def: Pick<ComponentDocApi, "moduleExports">) {
  return def.moduleExports
    .map((prop) => {
      const prop_comments = [addCommentLine(prop.description?.replace(/\n/g, "\n* "))].filter(Boolean).join("");

      let type_def = `export type ${prop.name} = ${prop.type || ANY_TYPE};`;

      const is_function = prop.type && /=>/.test(prop.type);

      if (is_function) {
        const [first, second, ...rest] = prop.type!.split("=>");
        const rest_type = rest.map((item) => "=>" + item).join("");

        type_def = `export declare function ${prop.name}${first}:${second}${rest_type};`;
      }

      return `
      ${prop_comments.length > 0 ? `/**\n${prop_comments}*/` : EMPTY_STR}
      ${type_def}`;
    })
    .join("\n");
}

export function writeTsDefinition(component: ComponentDocApi) {
  const {
    moduleName,
    typedefs,
    props,
    moduleExports,
    slots,
    events,
    rest_props,
    extends: _extends,
    componentComment,
  } = component;
  const { props_name, prop_def } = genPropDef({
    moduleName,
    props,
    rest_props,
    extends: _extends,
  });

  return `
  /// <reference types="svelte" />
  import type { SvelteComponentTyped } from "svelte";
  ${genImports({ extends: _extends })}
  ${genModuleExports({ moduleExports })}
  ${getTypeDefs({ typedefs })}
  ${prop_def}
  ${genComponentComment({ componentComment })}
  export default class ${moduleName === "default" ? "" : moduleName} extends SvelteComponentTyped<
      ${props_name},
      {${genEventDef({ events })}},
      {${genSlotDef({ slots })}}
    > {
      ${genAccessors({ props })}
    }`;
}

export interface WriteTsDefinitionsOptions {
  outDir: string;
  inputDir: string;
  preamble: string;
  exports: ParsedExports;
}

export default async function writeTsDefinitions(components: ComponentDocs, options: WriteTsDefinitionsOptions) {
  const ts_base_path = path.join(process.cwd(), options.outDir, "index.d.ts");
  const writer = new Writer({ parser: "typescript", printWidth: 80 });
  const indexDTs = options.preamble + createExports(options.exports, components);

  for await (const [moduleName, component] of components) {
    const ts_filepath = convertSvelteExt(path.join(options.outDir, component.filePath));
    await writer.write(ts_filepath, writeTsDefinition(component));
  }

  await writer.write(ts_base_path, indexDTs);

  console.log(`created TypeScript definitions.`);
}
