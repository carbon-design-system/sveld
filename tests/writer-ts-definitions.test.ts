import { test, expect, describe } from "vitest";
import { writeTsDefinition, formatTsProps, getTypeDefs } from "../src/writer/writer-ts-definitions";
import { ComponentDocApi } from "../src/rollup-plugin";
import { ParsedComponent } from "../src/ComponentParser";

describe("writerTsDefinition", () => {
  test("writeTsDefinition", () => {
    expect(formatTsProps(undefined)).toEqual("any");
    expect(formatTsProps("{ a: null }")).toEqual("{ a: null }\n");
    expect(getTypeDefs({ typedefs: [] })).toEqual("");
    expect(
      getTypeDefs({
        typedefs: [
          {
            type: "{ [key: string]: boolean; }",
            name: "MyTypedef",
            ts: "interface MyTypedef { [key: string]: boolean; }",
          },
        ],
      })
    ).toEqual("export interface MyTypedef { [key: string]: boolean; }");

    const parsed_output: ParsedComponent = {
      props: [
        {
          name: "propBool",
          kind: "let",
          type: "boolean",
          value: "true",
          isFunction: false,
          isFunctionDeclaration: false,
          constant: false,
          reactive: true,
        },
        {
          name: "propString",
          kind: "let",
          type: "string",
          value: '""',
          isFunction: false,
          isFunctionDeclaration: false,
          constant: false,
          reactive: false,
        },
        {
          name: "name",
          kind: "let",
          type: "string",
          isFunction: false,
          isFunctionDeclaration: false,
          constant: false,
          reactive: false,
        },
        {
          name: "id",
          kind: "let",
          type: "string",
          value: '"" + Math.random().toString(36)',
          isFunction: false,
          isFunctionDeclaration: false,
          constant: false,
          reactive: false,
        },
        {
          name: "propConst",
          kind: "const",
          type: "{ [key: string]: boolean; }",
          value: '{ ["1"]: true }',
          isFunction: false,
          isFunctionDeclaration: false,
          constant: true,
          reactive: false,
        },
        {
          name: "fn",
          kind: "function",
          type: "() => {     localBool = !localBool;   }",
          value: "() => {     localBool = !localBool;   }",
          isFunction: true,
          isFunctionDeclaration: false,
          constant: false,
          reactive: false,
        },
      ],
      moduleExports: [],
      slots: [
        {
          name: "__default__",
          default: true,
          fallback: "{name}",
          slot_props: "{}",
        },
      ],
      events: [],
      typedefs: [],
      rest_props: undefined,
    };

    const component_api: ComponentDocApi = {
      moduleName: "ModuleName",
      filePath: "./src/ModuleName.svelte",
      ...parsed_output,
    };

    expect(writeTsDefinition(component_api)).toMatchSnapshot();
  });

  test('"default" module name', () => {
    const component_api: ComponentDocApi = {
      moduleName: "default",
      filePath: "./src/ModuleName.svelte",
      props: [],
      moduleExports: [],
      slots: [],
      events: [],
      typedefs: [],
      rest_props: undefined,
    };

    expect(writeTsDefinition(component_api)).toMatchSnapshot();
  });
});
