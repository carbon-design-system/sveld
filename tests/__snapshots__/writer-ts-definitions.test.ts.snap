// Vitest Snapshot v1

exports[`writerTsDefinition > "default" module name 1`] = `
"
  /// <reference types=\\"svelte\\" />
  import { SvelteComponentTyped } from \\"svelte\\";
  
  
  
  
    export interface defaultProps  {
      
    }
  
  
  export default class  extends SvelteComponentTyped<
      defaultProps,
      {},
      {}
    > {
      
    }"
`;

exports[`writerTsDefinition > writeTsDefinition 1`] = `
"
  /// <reference types=\\"svelte\\" />
  import { SvelteComponentTyped } from \\"svelte\\";
  
  
  
  
    export interface ModuleNameProps  {
      
      /**
* @default true
*/
      propBool?: boolean;

      /**
* @default \\"\\"
*/
      propString?: string;

      /**
* @default undefined
*/
      name?: string;

      /**
* @default \\"\\" + Math.random().toString(36)
*/
      id?: string;

      /**
* @default () => { localBool = !localBool; }
*/
      fn?: () => {     localBool = !localBool;   };
    }
  
  
  export default class ModuleName extends SvelteComponentTyped<
      ModuleNameProps,
      {},
      {default: {}
;}
    > {
      
    
    propConst: { [key: string]: boolean; };
    }"
`;
